packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts

```diff
@@ -0,0 +1,305 @@
+import * as fs from 'fs';
+import * as path from 'path';
+import { Resource, ISynthesisSession, attachCustomSynthesis, Stack, Reference, Tokenization, IResolvable, StringConcat } from '@aws-cdk/core';
+import { Construct, Dependable, DependencyGroup } from 'constructs';
+import { Grant } from '../grant';
+import { IManagedPolicy } from '../managed-policy';
+import { Policy } from '../policy';
+import { PolicyDocument } from '../policy-document';
+import { PolicyStatement } from '../policy-statement';
+import { AddToPrincipalPolicyResult, IPrincipal, PrincipalPolicyFragment } from '../principals';
+import { IRole } from '../role';
+
+const POLICY_SYNTHESIZER_ID = 'PolicySynthesizer';
+
+/**
+ * Options for a precreated role
+ */
+export interface PrecreatedRoleProps {
+  /**
+   * The base role to use for the precreated role.
+   */
+  readonly role: IRole;
+
+  /**
+   * The assume role (trust) policy for the precreated role.
+   *
+   * @default - no assume role policy
+   */
+  readonly assumeRolePolicy?: PolicyDocument;
+
+  /**
+   * If the role is missing from the precreatedRole context
+   *
+   * @default false
+   */
+  readonly missing?: boolean;
+}
+
+/**
+ * An IAM role that has been created outside of CDK and can be
+ * used in place of a role that CDK _is_ creating.
+ *
+ * When any policy is attached to a precreated role the policy will be
+ * synthesized into a separate report and will _not_ be synthesized in
+ * the CloudFormation template.
+ */
+export class PrecreatedRole extends Resource implements IRole {
+  public readonly assumeRoleAction: string;
+  public readonly policyFragment: PrincipalPolicyFragment;
+  public readonly grantPrincipal = this;
+  public readonly principalAccount?: string;
+  public readonly roleArn: string;
+  public readonly roleName: string;
+  public readonly stack: Stack;
+
+  private readonly policySynthesizer: PolicySynthesizer;
+  private readonly policyStatements: string[] = [];
+  private readonly managedPolicies: string[] = [];
+
+  private readonly role: IRole;
+  constructor(scope: Construct, id: string, props: PrecreatedRoleProps) {
+    super(scope, id, {
+      account: props.role.env.account,
+      region: props.role.env.region,
+    });
+    this.role = props.role;
+    this.assumeRoleAction = this.role.assumeRoleAction;
+    this.policyFragment = this.role.policyFragment;
+    this.principalAccount = this.role.principalAccount;
+    this.roleArn = this.role.roleArn;
+    this.roleName = this.role.roleName;
+    this.stack = this.role.stack;
+
+    Dependable.implement(this, {
+      dependencyRoots: [this.role],
+    });
+
+    // add a single PolicySynthesizer under the `App` scope
+    this.policySynthesizer = (this.node.root.node.tryFindChild(POLICY_SYNTHESIZER_ID)
+      ?? new PolicySynthesizer(this.node.root)) as PolicySynthesizer;
+    this.policySynthesizer.addRole(this.node.path, {
+      roleName: props.missing ? 'missing role' : this.roleName,
```

Comment:
[nitpick] The goal here is to fill in the value. If the Role is created and the name is `missing role` it makes it a bit extra effort to troubleshoot/ audit.
Should it be a `Default`+timestamp?